### ISSUE TITLE
UnhandledPromiseRejection by using kraken.ws.openOrders({ token: token! }).subscribe()

### ISSUE THREAD

**Issue by aysuio** [jpcx/node-kraken-api#8]
Hi Justin, 

Thanks for the awesome work.

When i was using subscribe function from websocket, it basically return:

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "undefined".] {
  code: 'ERR_UNHANDLED_REJECTION'
}

Even adding catch() doesn't help on this case.

Any help would be appreciated

**Comment by jpcx** [comment]
Good catch, thank you! I added rejection handlers to several WS methods; can you update to 2.2.1 and try again?

**Comment by aysuio** [comment]
Hi Justin, 

I just loaded 2.2.1 and it still print me the same error. Here is a brutal code to reproduce the error

```
(async () => {
    const kraken = new Kraken()
    const pairs = await kraken.assetPairs()
    Object.values(pairs).forEach( pair => {
        const wspair = pair.wsname
        kraken.ws.spread().subscribe(wspair)
    })
})()
```

I ended up with the same error 

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected 
with the reason "undefined".] {
  code: 'ERR_UNHANDLED_REJECTION'
}

**Comment by aysuio** [comment]
Sorry for my bad coding skills. Not sure if I have done some dumb things, Thanks a lot if you could guide me through

**Comment by jpcx** [comment]
Ah don't worry! Thanks for bringing this to my attention. I found where the rejection was happening and fixed that, but I'm currently investigating why more than 49 simultaneous subscription requests causes the 50th to not receive a status and hang.

I had not noticed this before because my intention was for `subscribe` to be called with *all* desired pairs at once, because doing so creates only one outbound WS request.

I also needed to modify the `subscribe` signature to make it easier to input an array of strings; I had `(pair: string, ...pairs: string[])` to enforce at least one pair, but this makes subscribing with the output from `assetPairs` difficult.

I'll keep trying to troubleshoot the simultaneous request issue for a bit and will respond with an update soon.

Until I publish the updated `subscribe` function signature, you can use the code below:

```ts
import { Kraken } from "node-kraken-api";

(async () => {
  const kraken = new Kraken();
  kraken.ws.pub.on("error", console.error);
  const pairs = Object.values(await kraken.assetPairs())
    .map((x) => x.wsname)
    .filter((x) => x);
  const s: Kraken.WS.Spread.Subscriber = await (<any>(
    kraken.ws.spread().subscribe
  ))(...pairs);
  s.on("update", console.log);
  s.on("error", console.error);
})();
```

**Comment by aysuio** [comment]
XD.. I should have sub multiple pairs in one call. Thanks very much!

**Comment by jpcx** [comment]
No problem! I published 2.2.2 with the unhandled rejection fix and updated subscribe/unsubscribe signatures.

It turns out that some of the response errors were not being emitted, and making 50 or more simultaneous WS requests was triggering a server rate-limit error (so the 50th subscription was never resolving, but there was no outward indication of any errors).

I added an error handler for that case, but it's important to listen for general WS errors emitted by `ws.pub` and `ws.priv`.

Here is an updated snippet using the improved signature:

```ts
import { Kraken } from "node-kraken-api";

(async () => {
  const kraken = new Kraken();
  kraken.ws.pub.on("error", console.error);
  const pairs = <string[]>Object.values(await kraken.assetPairs())
    .map((x) => x.wsname)
    .filter((x) => x);
  await kraken.ws
    .spread()
    .on("update", console.log)
    .on("error", console.error)
    .subscribe(...pairs);
})();
```
